src/projection.jl

```diff
@@ -52,93 +128,197 @@ function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
 end
 
 # Generic
-(project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
 (::ProjectTo{T})(dx::T) where {T} = dx  # not always true, but we can special case for when it isn't
-(::ProjectTo{T})(dx::AbstractZero) where {T} = zero(T)
+(::ProjectTo{T})(dx::AbstractZero) where {T} = dx # zero(T)
+
+# Thunks
+(project::ProjectTo)(dx::Thunk) = Thunk(project ∘ dx.f)
+(project::ProjectTo)(dx::InplaceableThunk) = Thunk(project ∘ dx.val.f) # can't update in-place part
```

Comment:
I was messing with this a bit but don't really have any clarity. I wondered if instead of `dx -> dx .+= y` we should always write some `Fix1`-esque functions, `Free13(broadcast!, +, y)` or something, to make it easier to manipulate. Inserting `real` into the broadcast is the easy bit, dealing with `Diagonal` as you say seems harder, and for `mul!` maybe you just need a mechanism for deciding when to give up?

Other thoughts here: https://github.com/JuliaDiff/ChainRules.jl/pull/459#discussion_r664586841

Reply:
hmmm, yeah, not supporting `mul!` is bad. It's one of the most important use of InpaceableThunk.
(The other being `getindex`)
Maybe we just only write the one that uses an InplaceableThunk for the cases when it is a `Array{<:Number}`, which will not need projecting.
Since we only need the projection stuff for if dealing with AbstractArrays.

Reply:
Even for Arrays. If you do `z = x * y` with `x::Matrix{Float64}` and `y::Matrix{Complex{Float64}}`, that's already not BLAS, and Julia will (I think) now copy & promote `x`. In such cases, losing the `mul!` is no great tragedy. The trick is figuring out a mechanism to switch it off -- blindly trying `mul!(dx, ...)` will give errors. If you assume that `dz` has already been projected, then you can figure out whether you are safe or not from the types of `x` and `y` alone. But at present there's no machine to do so.

Reply:
This is easy if we just have a `rrule(*, ::Array{T}, ::Array{T}) where T` and maybe a few safe known unions around that (Probably `StridedArray`)
and that is the only one that uses a `InplaceThunk`.

and then we have another `rrule(*, ::AbstractArray, ::AbstractArray)` and that uses `project`.

And we keep `project` as the tool which you have to fix up rules that use AbstractArrays.
Rather than it being a thing you apply absolutely everywhere.

Reply:
Yes, that would work. It will double the length of `arraymath.jl`. I still wonder if it can't be automated a bit.

Maybe the in-place update itself should only ever apply to `dx::StridedArray` or something. It's hard to picture it really working well on structured types. That would solve the `dx::Diagonal` issue, and it remains to solve eltype. Perhaps you can pass a 3rd argument so `InplaceableThunk(þ, f!, T)` is only applied to `dx::StridedArray{<:T}`, otherwise it falls back to something like `dx .+= Project{eltype(dx)}.(þ)`. Where ideally you figure out a way to fuse that broadcast with the one in the thunk, if there is one.

Reply:
This topic should move to a new issue, IMO. It's fairly separate and surely isn't going to be solved in this PR. Once can't magically transfer threads, I guess?